src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -34,8 +34,100 @@
 #include "runtime/vm_version.hpp"
 #include "utilities/powerOfTwo.hpp"
 
+class AsmGenerator {
+public:
+
+  class WrapperNode: public ResourceObj {
+  public:
+    WrapperNode * _next;
+    virtual void operator()() = 0;
+  };
+
+  template <typename T>
+  class LambdaWrapper : public WrapperNode {
+  public:
+    T _t;
+    LambdaWrapper(T t): _t(t) {}
+    virtual void operator()() {
+      _t();
+    }
+  };
+
+  WrapperNode *_holder_list_head, *_last;
+  int _count;
+
+  AsmGenerator()
+    : _holder_list_head(nullptr), _last(nullptr), _count(0) {}
+
+  WrapperNode *last() {
+    return _last;
+  }
+
+  template<typename T>
```

Comment:
This probably also needs a comment.

    /*
     * Appends a closure to the generator that can be executed
     * later to append one or more instructions to the target code
     * buffer. Normally each closure generates a single instruction,
     * allowing multiple generators that interleave parallel instruction
     * streams to obtain the maximum opportunities for pipeline
     * parallelism. In cases where an individual instruction stream
     * uses a shared scratch register to hold a temporary that will be
     * consumed by a later instruction the closure must generate the
     * full instruction sequence between the writer and last reader of
     * the temporary as a block, ensuring that instructions from
     * parallel streams which write or read the same scratch register
     * cannot side-affect each other.
     */